OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs

```diff
@@ -191,15 +191,24 @@ bool CycleBases()
 				.Select(b => b.Actor)
 				.ToList();
 
+			// If no base is found all existing buildings are selected.
 			if (!bases.Any())
-				return true;
+			{
+				bases = world.ActorsWithTrait<TargetableBuilding>()
+				.Where(a => a.Actor.Owner == world.LocalPlayer)
+				.Select(b => b.Actor)
+				.ToList();
+				if (!bases.Any())
+					return true;
+			}
 
 			var next = bases
 				.SkipWhile(b => !world.Selection.Actors.Contains(b))
 				.Skip(1)
 				.FirstOrDefault();
 
-			if (next == null)
+			// If no construction yard exists then next is set to only the oldest existing building
+			if (next == null || !next.HasTrait<BaseBuilding>())
```

Comment:
Without that check using the hotkey will cycle through all buildings.  The check results in going to the oldest existing building which is more inline with how Red Alert handles returning to base without a construction yard.

---- On Sat, 04 Jul 2015 15:01:39 -0600 Pavel Penev &lt;notifications@github.com&gt; wrote ---- 

In OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs:
 &gt; &gt; var next = bases &gt; .SkipWhile(b =&gt; !world.Selection.Actors.Contains(b)) &gt; .Skip(1) &gt; .FirstOrDefault(); &gt; &gt; - if (next == null) &gt; + // If no construction yard exists then next is set to only the oldest existing building &gt; + if (next == null || !next.HasTrait&lt;BaseBuilding&gt;())  What is the purpose of this additional check for BaseBuilding?
 —
Reply to this email directly or view it on GitHub.
    
 
 




​